src/nn/nn_minimax_strategy.py

```diff
@@ -1,57 +1,74 @@
+        def fn_eval(game, alpha, beta):
+            return self._value_with_lookahead(game, player_id, self.lookahead_limit, alpha, beta)
 
         t0 = time.time()
-        move = self._find_best_move_and_score_for_player(game, player_id, fn_eval)[0]
+        move = self._find_best_move_and_score_for_player(game, player_id, fn_eval, -math.inf, math.inf)[0]
         self.total_time += time.time() - t0
         self.total_moves += 1
         return move
 
     def average_move_time(self):
         return self.total_time / self.total_moves
 
-    def _value_with_lookahead(self, game, player_id, lookaheads_remaining):
+    def _value_with_lookahead(self, game, curr_player, lookaheads_remaining, alpha, beta):
         if lookaheads_remaining == 0:
-            board_state = self._build_board_state(game, player_id)
+            board_state = self._build_board_state(game, curr_player)
             score = self.network.eval_position(board_state)
         else:
-            other_player_id = self._get_other_player(player_id)
-            def fn_eval(game):
-                return self._value_with_lookahead(game, other_player_id, lookaheads_remaining-1)
-            score = -self._find_best_move_and_score_for_player(game, other_player_id, fn_eval)[1]
+            other_player_id = self._get_other_player(curr_player)
+
+            def fn_eval(game, alpha, beta):
+                return self._value_with_lookahead(game, other_player_id, lookaheads_remaining-1, alpha, beta)
+            score = -self._find_best_move_and_score_for_player(
+                    game, other_player_id, fn_eval, alpha, beta)[1]
 
         return score
 
-    def _find_best_move_and_score_for_player(self, game, player_id, fn_eval):
+    def _find_best_move_and_score_for_player(self, game, curr_player, fn_eval, alpha, beta):
         valid_moves = game.board.get_valid_moves()
         move_scores = {}
         for move in valid_moves:
             game_copy = game.clone()
-            game_copy.move(move, player_id)
+            game_copy.move(move, curr_player)
             if game_copy.finished:
-                move_scores[move] = math.inf if game_copy.winner == player_id else 0 if not game_copy.winner else -math.inf
+                # This tricky line sets the score to be +inf if the curr_player (could be player_id or opponent) wins
+                # Score is 0 on a tie and -inf if the curr_player loses
+                move_scores[move] = math.inf if game_copy.winner == curr_player else 0 if not game_copy.winner else -math.inf
             else:
-                move_scores[move] = fn_eval(game_copy)
+                # Because we are using the Nagamax variation of minimax, we must negate and flip our alpha & beta
+                # values every time we recurse to look ahead one more.
+                # see more here: https://en.wikipedia.org/wiki/Negamax
+                move_scores[move] = fn_eval(game_copy, -beta, -alpha)
+
+            if alpha < move_scores[move]:
```

Comment:
No, because the sign flips are key to alpha beta pruning with the Negamax algorithm. In the case where scores are negative, as often happens with odd numbered look aheads, that negative score is how badly our opponent's best move will hurt us. So by picking the max, among negative values, we are ensuring the minimize damage to our standing, and maximize our own standing. Does that make sense?